dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java

```diff
@@ -70,7 +70,9 @@
 
     private static int DEFAULT_BUFFER_SIZE = 10_000;
     private final Predicate<C> joinCondition;
-    private final Multimap<Integer, Object[]> buffer;
+    private final com.carrotsearch.hppc.IntObjectHashMap<Object[]> buffer;
```

Comment:
Why are we creating 2 maps? Wouldn't one be enough?

Reply:
It's because we wanted to avoid having the List created when there are no collisions (in which case the `buffer` would hold the row). It's a memory usage optimisation

Reply:
If there are duplicate rows it would still fallback to the second map. This optimization seems to be very data dependent. Are you certain that we're doing the right thing here?

Reply:
@mfussenegger Of course, all of those have to do with the nature of the data. My thinking is that it quite often PKs are used for join conditions thus we would instantiate lists for every row which is both cpu consuming + more memory needed. Of course if we end up with duplication for all/most values then the idea of the 2 buffers actually performs worst...